### ISSUE TITLE
pkg/storage/redis: connection failure handling

### ISSUE THREAD

**Issue by travisgroth** [pomerium/pomerium#1167]
## What happened?

When connectivity to redis is severed, databroker goes into unrecoverable state, even if redis comes back at the same address.

I believe health checks are still passing but I need to confirm in dedicated service mode.  This was tested in all-in-one.

## What did you expect to happen?

We have a few options:

- Easiest is probably to bubble up the failure and exit so that monitoring/automation can try to handle recovery
- Ideally, we'd respond with errors to RPCs and reconnect with backoff.  This would allow for transient failure, failover, DNS remapping, etc.
- Other ideas?

## How'd it happen?

1. Ran with redis databroker backend
2. Restarted redis
3. Infinitely scrolling errors in logs and no recovery

## What's your environment like?

- Pomerium version (retrieve with `pomerium --version` or `/ping` endpoint): 62f8348d0e4bc238df368027d0b18fb734953f26 (https://github.com/pomerium/pomerium/pull/1163)
- Server Operating System/Architecture/Cloud: Local/OSX

## What did you see in the logs?

This repeats at a high rate forever:
```logs
{"level":"error","error":"EOF","time":"2020-07-30T14:36:17-04:00","message":"failed to receive from redis channel"}
```

**Comment by cuonglm** [comment]
I think this is partly fixed in #1163.

The problem is redigo does not return net.Err, so pub sub still uses old conn.

**Comment by travisgroth** [comment]
>I think this is partly fixed in #1163.

This was actually on that branch

**Comment by cuonglm** [comment]
> >I think this is partly fixed in #1163.
> 
> 
> 
> This was actually on that branch

That's why I said partly. I think we should always return if got error, then the client will re-call Sync to recover
